app-typescript/components/TreeSelect/TreeSelect.tsx

```diff
@@ -671,38 +701,98 @@ export class TreeSelect<T> extends React.Component<IProps<T>, IState<T>> {
                                     <i className="icon-plus-large"></i>
                                 </button>
                             }
+                            {this.props.dragAndDrop
+                                ? (
+                                    <DragDropContext onDragEnd={this.onDragEnd}>
+                                        <Droppable droppableId="droppable" direction="horizontal">
+                                            {(provided, _snapshot) => (
+                                                <ul
+                                                    className="tags-input__tag-list"
+                                                    ref={provided.innerRef}
+                                                    {...provided.droppableProps}
+                                                >
+                                                    {this.state.value.map((item, i: number) => {
+                                                        const Wrapper: React.ComponentType<{backgroundColor?: string}>
+                                                        = ({backgroundColor, children}) => (
+                                                            <TreeSelectPill
+                                                                item={item}
+                                                                readOnly={this.props.readOnly}
+                                                                disabled={this.props.disabled}
+                                                                valueTemplate={this.props.valueTemplate}
+                                                                backgroundColor={backgroundColor}
+                                                                onRemove={() => this.removeClick(i)}
+                                                                getBackgroundColor={this.props.getBackgroundColor}
+                                                                draggable={this.props.dragAndDrop}
+                                                            >
+                                                                {children}
+                                                            </TreeSelectPill>
+                                                        );
+
+                                                        return (
+                                                            <Draggable key={i} draggableId={`${i}`} index={i}>
```

Comment:
when it's possible for items to change order, item index in the array can not be used as a react key